### ISSUE TITLE
Try to repair documents with invalid XREF table

### ISSUE THREAD

**Issue by gtoffoli** [jorisschellekens/borb#81]
I've posted on StackOverflow the issue "What are known limitations of borb related to PDF versions?" and got a timely answer. Thanks.
Now, I have better focused a problem.

I repeatedly get exceptions such as the AssertionError on "assert isinstance(xref_stream, Stream)".
I realized that
- this occurs systematically with documents that presumably were generated by exporting as PDF articles being distributed by the ProQuest information-content service (proquest.com)
- these documents are the same on which also PyPDF4 goes in error ("Could not find xref table at specified location"); in fact, I'm trying borb on the assumption that PyPDF4 isn't well maintained.

The same documents are visualized without problems by Adobe Acrobat and by the internal PDF visualizers of FireFox and Chrome. Moreover, on StackOverflow I read that most times also pikepdf (based on qpdf) is able to repair invalid PDF documents suffering of similar problems.

Would it be very hard trying to make also borb resilient in this respect?

Thank you in advance in any case.

I cannot attach a sample here, since I got those documents privately.
In the document properties, from the menu of Adobe Acrobat, I see empty meta-data fields (only dates) and: PDF author = "iText 2.1.7 by 1T3XT", PDF version = "1.4 (Acrobat 5.x)".

**Comment by jorisschellekens** [comment]
Hi there,

I think fixing a broken xref might be a lot more tricky than expected.

If the location of the xref is wrong, it seems logica to assume that all byte offsets will be wrong. 

That means scanning the entire document for object declarations, and trying to piece back an xref from that.

I'll see if I can't create such a document, to at least get an idea of the scope of the problem.

Kind regards,
Joris Schellekens

**Comment by jorisschellekens** [comment]
I think I've found a way to fix the `xref` table.

```python
import io
import logging
import typing

from borb.io.read.tokenize.high_level_tokenizer import HighLevelTokenizer
from borb.io.read.types import Reference, Name
from borb.pdf.xref.plaintext_xref import PlainTextXREF
from borb.pdf.xref.xref import XREF

logger = logging.getLogger(__name__)


class RebuiltXREF(PlainTextXREF):

    def read(
        self,
        src: typing.Union[io.BufferedIOBase, io.RawIOBase, io.BytesIO],
        tok: HighLevelTokenizer,
        initial_offset: typing.Optional[int] = None,
    ) -> "XREF":

        # get pos
        pos_before: int = src.tell()

        # set pos
        src.seek(0)

        # get object declarations
        i: int = 0
        trailer_pos: typing.Optional[int] = None
        bytes_in_pdf: bytes = src.read()
        while i < len(bytes_in_pdf):
            # 0 0 obj
            if (
                48 <= bytes_in_pdf[i] <= 57             # <number>
                and bytes_in_pdf[i + 1] == 32           # <space>
                and 48 <= bytes_in_pdf[i + 2] <= 57     # <number>
                and bytes_in_pdf[i + 3] == 32           # <space>
                and bytes_in_pdf[i + 4] == 111          # 'o'
                and bytes_in_pdf[i + 5] == 98           # 'b'
                and bytes_in_pdf[i + 6] == 106          # 'j'
            ):
                logger.debug("%d %d obj at %d" % (bytes_in_pdf[i] - 48, bytes_in_pdf[i+1] - 48, i))
                self._entries.append(
                    Reference(
                        object_number=bytes_in_pdf[i] - 48,
                        generation_number=bytes_in_pdf[i + 2] - 48,
                        byte_offset=i,
                    )
                )
                i += 7
                continue
            # 00 0 obj
            if (
                48 <= bytes_in_pdf[i] <= 57             # <number>
                and 48 <= bytes_in_pdf[i+1] <= 57       # <number>
                and bytes_in_pdf[i + 2] == 32           # <space>
                and 48 <= bytes_in_pdf[i + 3] <= 57     # <number>
                and bytes_in_pdf[i + 4] == 32           # <space>
                and bytes_in_pdf[i + 5] == 111          # 'o'
                and bytes_in_pdf[i + 6] == 98           # 'b'
                and bytes_in_pdf[i + 7] == 106          # 'j'
            ):
                logger.debug("%d %d obj at %d" % ((bytes_in_pdf[i] - 48) * 10 + (bytes_in_pdf[i+1] - 48), bytes_in_pdf[i+3] - 48, i))
                self._entries.append(
                    Reference(
                        object_number=(bytes_in_pdf[i] - 48) * 10 + (bytes_in_pdf[i+1] - 48),
                        generation_number=bytes_in_pdf[i + 3] - 48,
                        byte_offset=i,
                    )
                )
                i += 8
                continue

            # 000 0 obj
            # TODO

            # trailer
            if (
                bytes_in_pdf[i] == 116          # 't'
                and bytes_in_pdf[i + 1] == 114  # 'r'
                and bytes_in_pdf[i + 2] == 97   # 'a'
                and bytes_in_pdf[i + 3] == 105  # 'i'
                and bytes_in_pdf[i + 4] == 108  # 'l'
                and bytes_in_pdf[i + 5] == 101  # 'e'
                and bytes_in_pdf[i + 6] == 114  # 'r'
            ):
                trailer_pos = i

            # default
            i += 1

        # read trailer
        src.seek(trailer_pos)
        self[Name("Trailer")] = self._read_trailer(src, tok)
        self[Name("Trailer")].set_parent(self)

        # restore pos
        src.seek(pos_before)

        # return
        return self

```

I'm going to add that as an alternative parser for `xref` tables (alongside the existing parser for plaintext `xref` and streaming `xref`).

I have a test-case now that corrupts a `PDF`, so I can check fairly easily whether the new parser holds up.

I also took the liberty of answering your StackOverflow post.

Kind regards,
Joris Schellekens

**Comment by jorisschellekens** [comment]
I'm currently running all the tests (usually takes an hour or two). After that I need to write the release-notes. And then I can release version `2.0.21` which should offer some support for rebuilding `xref`.

**Comment by jorisschellekens** [comment]
Hi,

Version `2.0.21` is now available.
The aforementioned test can be found [here](https://github.com/jorisschellekens/borb/blob/master/tests/pdf/document/fix_broken/test_fix_broken_xref.py)

You did not share your document, so I can't know for sure whether I have fixed the problem for your document in particular. But I have tested this against my demo-document, and against the larger corpus of PDF documents I keep [here](https://github.com/jorisschellekens/pdf-corpus). The "being able to open" rate jumped from 88% to 100%, so it certainly seems to have had some effect.

Thank you for the awesome suggestion! This little code-snippet has made a tremendous impact.

Kind regards,
Joris Schellekens

**Comment by gtoffoli** [comment]
Hi, only now I was able to deal again with the issue I had pointed out and, after some painstaking work, I think I've solved the puzzle!
I've analyzed in detail only one of the sample PDF documents that I was mentioning, but optimistically I assume that the others suffer from the same problem.
I attach it to this post.
[ProQuest1.pdf](https://github.com/jorisschellekens/borb/files/8315147/ProQuest1.pdf)

1. SOME FACTS
The size of the PDF document is 197301 bytes.
It has 2 xrefs. The last xref has 50 items; an intermediate xref has 37 items.

2. THE RESULT OF LOADING AND DUMPING THE FILE WITH BORB
The very small ouput file has one xref with only 32 items (neither 50, nor 87, nor 85: see below).
Opening it with:
- Acrobat Reader: ".. opening document .. this file cannot be open since it contains no page".
- FireFox: opens, shows empty view, empty outline.
- Chrome: opens, shows title, empty view, empty outline.

3. MY CONCLUSION IN BRIEF
Apparently, the last xref has an inconsistent startxref offset, while the other one has a consistent offset.
If, with a text editor, I truncate the file after the %%EOF line of the first trailer, I get a valid document with (almost) the expected content!
If I remove the first part of the file, up to and including the %%EOF line of the trailer associated to the first xref, the remainder becomes consistent and I get a valid document with the expected content! In the first case the page footer, including an image, is missing.

4. SOME OTHER FINDINGS
The trailer of the last xref doesn't have a prev/Prev/Previous key; thus, borb tries to load only the last one. After the last trailer, at offset 197278, finds 'startxref' and reads the xref offset: 105412
But at offset 105412 it finds a /Subtype keyword in place of 'xref', which is at offset 196150; so, on reading the xref with PlainTextXREF, an assertion fails and RebuiltXREF is the last resort.
At the end of the execution of _read_xref,  with RebuiltXREF,  it seems that len(most_recent_xref) = 85. Maybe, it isn't a coincidence that  85 is the result of the sum: 50 (size of last xref) + 37 (size of the other xref) - 2. I subtract 2 since the count 37 includes a free object (the first one) and a duplicate one (ID 00000015), that is one present also in the last xref.

I understand that brutally removing part of the file is not a general solution.
Possibly, before reconstructing the xref from the content, it could be worth trying to correct the wrong offset of the last xref, by subtracting the size of the document up to the preceding EOF.
